File: palindromPartition.java
import java.util.*;
import java.io.*;
import  java.util.stream.*;
public class palindromPartition{
	public static void main(String [] args){
		String str = "aab";
		allCombination(str);
	}
	static void allCombination(String input){
		int n=input.length();
		List<List<String>> allpart = new ArrayList<>();
		Deque<String> curr = new LinkedList<>();
		allCombinationUtil(allpart,curr,0,n,input);
		System.out.println(allpart);

	}
	static void allCombinationUtil(List<List<String>> allPart,Deque<String> curr,int start,int end,String input){
		if(start>=end){
			allPart.add(new ArrayList<>(curr));
			return;
		}
		for(int i=start;i<end;i++){
			if(isPalindrom(input,start,i)){
				curr.addLast(input.substring(start,i+1));
				allCombinationUtil(allPart,curr,i+1,end,input);
				curr.removeLast();
			}
		}
	}
	static boolean isPalindrom(String str,int i,int j){
		if(i>j) return true;
		if(str.charAt(i)!=str.charAt(j)) return false;

		return isPalindrom(str,i+1,j-1);
	}
}

File: palindromPartition.cpp
#include<bits/stdc++.h>
using namespace std;
bool isPalindrom(string str,int i,int j){
    if(i>j)return true;
    else if(str.at(i)!=str.at(j)) return false;
    return isPalindrom(str,i+1,j-1);
}
void palindromPartition(string str,int start,int end,vector<string> &curr,vector<vector<string>> &res){
    if(start>=end){
        res.push_back(curr);
        return;
    }
    for(int i=start;i<end;i++){
        if(isPalindrom(str,start,i)){
            // cout<<str.substr(start,i+1-start)<<endl;
            curr.push_back(str.substr(start,i+1-start));
            palindromPartition(str,i+1,end,curr,res);
            curr.erase(curr.end());
        }
    }
}
int main(){
    string str = "aab";
    int length = str.length();
    vector<string> curr;
    vector<vector<string>> res ;
    palindromPartition(str,0,length,curr,res);
    for(auto i=0;i< res.size();i++){
        for(auto j = res[i].begin();j!=res[i].end();++j){
            cout<< *j <<" ";
        }
        cout<<endl;
    }
}